### PR TITLE
feat: Object store persistence in operator

### PR DESCRIPTION
**Pull request by dmartinol**
## What this PR does / why we need it:
This PR adds object store persistence configuration options to the FeatureStore CRD.
- Support is provided only to the Registry service.
- Additional S3 configurations are also supported
- Follows definitions in [RFC #42](https://docs.google.com/document/d/1vGKMizf3_14IyiF_W_Ik7CR03joFkQfzbKT0jH4PZJM/edit?tab=t.0#heading=h.9gaqqtox9jg6)
- Configuration examples are provided in config/samples

```yaml
apiVersion: feast.dev/v1alpha1
kind: FeatureStore
metadata:
  name: sample-s3-registry
spec:
  feastProject: my_project
  services:
    onlineStore:
      persistence:
        file:
          path: /data/online_store.db
    offlineStore:
      persistence:
        file:
          type: dask
    registry:
      local:
        persistence:
          file:
            path: s3://bucket/registry.db
            s3_additional_kwargs:
              ServerSideEncryption: AES256
              ACL: bucket-owner-full-control
              CacheControl: max-age=3600
```
## Which issue(s) this PR fixes:
Relates to https://github.com/feast-dev/feast/issues/4561

### CODE REVIEW

**Comment by dmartinol** [comment]
@tchughesiv @redhatHameed FYI

File: infra/feast-operator/api/v1alpha1/featurestore_types.go [hunk 1]
```diff
@@ -122,11 +123,14 @@ type RegistryPersistence struct {
 }
 
 // RegistryFilePersistence configures the file-based persistence for the registry service
-// +kubebuilder:validation:XValidation:rule="(!has(self.pvc) && has(self.path)) ? self.path.startsWith('/') : true",message="Ephemeral stores must have absolute paths."
+// +kubebuilder:validation:XValidation:rule="(!has(self.pvc) && has(self.path)) ? (self.path.startsWith('/') || self.path.startsWith('s3://') || self.path.startsWith('gs://')) : true",message="Ephemeral stores must use absolute paths or be S3 ('s3://') or GS ('gs://') object store URIs."
```
Comment (lokeshrangineni) [inline infra/feast-operator/api/v1alpha1/featurestore_types.go:126]: Do we need to use ephemeral when we are referring the s3 or gs bucket?

File: infra/feast-operator/api/v1alpha1/featurestore_types.go [hunk 2]
```diff
@@ -122,11 +123,14 @@ type RegistryPersistence struct {
...
 // +kubebuilder:validation:XValidation:rule="(has(self.pvc) && has(self.path)) ? !self.path.startsWith('/') : true",message="PVC path must be a file name only, with no slashes."
+// +kubebuilder:validation:XValidation:rule="(has(self.pvc) && has(self.path)) ? !(self.path.startsWith('s3://') || self.path.startsWith('gs://')) : true",message="PVC persistence does not support S3 or GS object store URIs."
+// +kubebuilder:validation:XValidation:rule="(has(self.s3_additional_kwargs) && has(self.path)) ? self.path.startsWith('s3://') : true",message="Additional S3 settings are available only for S3 object store URIs."
 type RegistryFilePersistence struct {
-	Path      string     `json:"path,omitempty"`
-	PvcConfig *PvcConfig `json:"pvc,omitempty"`
+	Path          string             `json:"path,omitempty"`
+	PvcConfig     *PvcConfig         `json:"pvc,omitempty"`
+	S3AddtlKwargs *map[string]string `json:"s3_additional_kwargs,omitempty"`
```
Comment (lokeshrangineni) [inline infra/feast-operator/api/v1alpha1/featurestore_types.go:133]: my personal preference to use full name `S3AdditionalKwargs` especially when only it is short of 4 chars. Feel free to decide.

**Review by redhatHameed — APPROVED** [review]
lgtm

File: infra/feast-operator/api/v1alpha1/zz_generated.deepcopy.go [hunk 3]
```diff
@@ -452,6 +452,17 @@ func (in *RegistryFilePersistence) DeepCopyInto(out *RegistryFilePersistence) {
 		*out = new(PvcConfig)
 		(*in).DeepCopyInto(*out)
 	}
+	if in.S3AddtlKwargs != nil {
+		in, out := &in.S3AddtlKwargs, &out.S3AddtlKwargs
+		*out = new(map[string]string)
```
Comment (lokeshrangineni) [inline infra/feast-operator/api/v1alpha1/zz_generated.deepcopy.go:457]: this has been reassigned in the deep copy using make. Do we need to initialize with empty map?

**Review by tchughesiv — APPROVED** [review]
lgtm

File: infra/feast-operator/test/api/featurestore_types_test.go [hunk 4]
```diff
@@ -111,6 +146,77 @@ func registryWithRelativePathForEphemeral(featureStore *feastdevv1alpha1.Feature
...
+	return copy
+}
+func registryWithGsBucketForPvc(featureStore *feastdevv1alpha1.FeatureStore) *feastdevv1alpha1.FeatureStore {
+	copy := featureStore.DeepCopy()
+	copy.Spec.Services = &feastdevv1alpha1.FeatureStoreServices{
+		Registry: &feastdevv1alpha1.Registry{
+			Local: &feastdevv1alpha1.LocalRegistryConfig{
+				Persistence: &feastdevv1alpha1.RegistryPersistence{
+					FilePersistence: &feastdevv1alpha1.RegistryFilePersistence{
+						Path: "gs://bucket/registry.db",
```
Comment (lokeshrangineni) [inline infra/feast-operator/test/api/featurestore_types_test.go:175]: above method `registryWithS3BucketForPvc` and this method seems to be the same except difference in the path. can we just pass that as a method param?

File: infra/feast-operator/test/api/featurestore_types_test.go [hunk 5]
```diff
@@ -67,6 +67,41 @@ func onlineStoreWithRelativePathForEphemeral(featureStore *feastdevv1alpha1.Feat
...
+	}
+	return copy
+}
+func onlineStoreWithGsBucketForPvc(featureStore *feastdevv1alpha1.FeatureStore) *feastdevv1alpha1.FeatureStore {
+	copy := featureStore.DeepCopy()
+	copy.Spec.Services = &feastdevv1alpha1.FeatureStoreServices{
+		OnlineStore: &feastdevv1alpha1.OnlineStore{
+			Persistence: &feastdevv1alpha1.OnlineStorePersistence{
+				FilePersistence: &feastdevv1alpha1.OnlineStoreFilePersistence{
+					Path: "gs://bucket/online_store.db",
```
Comment (lokeshrangineni) [inline infra/feast-operator/test/api/featurestore_types_test.go:93]: above method `onlineStoreWithS3BucketForPvc` this one appears to be the same except the difference in the path. can we pass as method param?

File: infra/feast-operator/api/v1alpha1/featurestore_types.go (hunk 1 above)
Comment (dmartinol) [reply infra/feast-operator/api/v1alpha1/featurestore_types.go:126]: you are right, trying a more generic `Registry files`, thanks

File: infra/feast-operator/api/v1alpha1/zz_generated.deepcopy.go (hunk 3 above)
Comment (dmartinol) [reply infra/feast-operator/api/v1alpha1/zz_generated.deepcopy.go:457]: this file is generated by `make generate`, I assume it is corrrect 😉

**Review by lokeshrangineni — APPROVED** [review]

**Pull request merged**